### PR TITLE
W-12075637: Separate modules in well defined levels for the tracing configurations

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoExportTillExecuteNextComponentCoreInfoProviderStart.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.api.tracing.customization;
+
+import org.mule.runtime.api.component.Component;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.tracer.api.span.info.StartSpanInfo;
+
+public class NoExportTillExecuteNextComponentCoreInfoProviderStart extends ComponentCoreStartSpanInfoProvider {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoExportTillExecuteNextComponentCoreInfoProviderStart.java:14]: Is the "Start" correctly placed?

File: core/src/main/java/org/mule/runtime/core/api/tracing/customization/FixedNameStartSpanInfo.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.api.tracing.customization;
+
+import org.mule.runtime.tracer.api.span.info.StartExportInfo;
+import org.mule.runtime.tracer.api.span.info.StartSpanInfo;
+
+/**
+ * A {@link StartSpanInfo} with a fix name.
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/api/tracing/customization/FixedNameStartSpanInfo.java:14]: fixed*

File: core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoExportableNoExportableFixedNameCoreSpanCustomizationInfoProvider.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+package org.mule.runtime.core.api.tracing.customization;
+
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.tracer.api.span.info.StartSpanInfo;
+
+public class NoExportableNoExportableFixedNameCoreSpanCustomizationInfoProvider
+    extends FixedNameCoreSpanCustomizationInfoProvider {
+
+  public NoExportableNoExportableFixedNameCoreSpanCustomizationInfoProvider(String name) {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoExportableNoExportableFixedNameCoreSpanCustomizationInfoProvider.java:16]: "NoExportable" is duplicated right?

File: core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoFixedNameSpanCustomizationInfo.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.api.tracing.customization;
+
+import org.mule.runtime.tracer.api.span.info.StartExportInfo;
+
+public class NoFixedNameSpanCustomizationInfo extends FixedNameStartSpanInfo {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/api/tracing/customization/NoFixedNameSpanCustomizationInfo.java:12]: Could we add a Javadoc to this class?

File: modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/DecoratedMuleOpenTelemetrySpan.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+import static io.opentelemetry.api.trace.StatusCode.ERROR;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import org.mule.runtime.api.profiling.tracing.SpanError;
```
Comment (IvanAndresFritzler) [inline modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/DecoratedMuleOpenTelemetrySpan.java:26]: Order of the imports

File: modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/NoopMuleOpenTelemetrySpan.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.context.Context;
+import org.mule.runtime.tracer.api.span.InternalSpan;
```
Comment (IvanAndresFritzler) [inline modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/NoopMuleOpenTelemetrySpan.java:21]: Order of the imports

File: modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/OpenTelemetryResources.java [hunk 7]
```diff
@@ -5,62 +5,53 @@
...
 import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
 import io.opentelemetry.sdk.trace.SdkTracerProviderBuilder;
 import io.opentelemetry.sdk.trace.SpanProcessor;
-import io.opentelemetry.sdk.trace.SdkTracerProvider;
 import io.opentelemetry.sdk.trace.data.SpanData;
 import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
 import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;
 import io.opentelemetry.sdk.trace.export.SpanExporter;
+import org.mule.runtime.tracer.api.sniffer.ExportedSpanSniffer;
```
Comment (IvanAndresFritzler) [inline modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/OpenTelemetryResources.java:27]: Order of the imports.

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

**Review by IvanAndresFritzler — DISMISSED** [review]

**Review by IvanAndresFritzler — APPROVED** [review]

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java [hunk 8]
```diff
@@ -526,25 +528,24 @@ private void beforeComponentProcessingStrategy(Processor processor, String proce
...
-      spanCustomizationInfo =
-          new NamedSpanBasedOnComponentIdentifierAloneSpanCustomizationInfo((Component) processor,
-                                                                            chainSpanCustomizationInfo.isPolicySpan());
+      startSpanInfo =
+          new ComponentEventBasedStartSpanInfoProvider((Component) processor).get(event);
     } else {
-      // Other type of processors we define taking into account the parent.
-      spanCustomizationInfo = getCustomizationInfo(chainSpanCustomizationInfo.isPolicySpan());
+      // Other processors are not exported
+      startSpanInfo = new NoExportFixedNameEventBasedStartSpanInfoProvider("unknown").get(event);
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java]: Extract "unknown" to constant indicating what it stands for

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java [hunk 9]
```diff
@@ -117,8 +118,8 @@ protected MessageProcessorChain createSimpleChain(List<Processor> tempList,
                                                     Optional<ProcessingStrategy> processingStrategyOptional) {
     if (tempList.size() == 1 && tempList.get(0) instanceof DefaultMessageProcessorChain) {
       DefaultMessageProcessorChain messageProcessorChain = (DefaultMessageProcessorChain) tempList.get(0);
-      if (spanCustomizationInfo != null) {
-        messageProcessorChain.setSpanCustomizationInfo(spanCustomizationInfo);
+      if (eventBasedStartStartInfoProvider != null) {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java:121]: Start is duplicated -> eventBasedStartInfoProvider*

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java [hunk 10]
```diff
@@ -366,8 +367,9 @@ public void setMessagingExceptionHandler(FlowExceptionHandler messagingException
     }
 
     @Override
-    public void setSpanCustomizationInfo(SpanCustomizationInfo spanCustomizationInfo) {
-      this.spanCustomizationInfo = spanCustomizationInfo;
+    public void setCoreSpanCustomizationInfoProvider(
+                                                     EventBasedStartSpanInfoProvider eventBasedStartStartInfoProvider) {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java]: eventBasedStartInfoProvider*

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/MessageProcessorChainBuilder.java [hunk 11]
```diff
@@ -59,7 +59,7 @@ public interface MessageProcessorChainBuilder extends MessageProcessorBuilder {
    *
    * @since 4.5.0
    */
-  default void setSpanCustomizationInfo(SpanCustomizationInfo spanCustomizationInfo) {
+  default void setCoreSpanCustomizationInfoProvider(EventBasedStartSpanInfoProvider spanCustomizationInfo) {
```
Comment (sofiamorseletto) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/MessageProcessorChainBuilder.java]: Could we change spanCustomizationInfo -> spanCustomizationInfoProvider ?

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessful.java [hunk 12]
```diff
@@ -42,7 +42,8 @@ public void initialise() throws InitialisationException {
     Long routeNumber = 1L;
     for (ProcessorRoute route : routes) {
       route.setMessagingExceptionHandler(null);
-      route.setSpanCustomizationInfo(new NamedSpanBasedOnComponentIdentifierAndAttemptSpanCustomizationInfo(this, routeNumber));
+      route.setCoreSpanCustominzationInfoProvider(new SuffixComponentEventBasedStartSpanInfoProvider(this,
```
Comment (sofiamorseletto) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessful.java]: setCoreSpanCustomizationInfoProvider*

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/ProcessorRoute.java [hunk 13]
```diff
@@ -72,7 +74,7 @@ public void setMessagingExceptionHandler(FlowExceptionHandler messagingException
     this.messagingExceptionHandler = messagingExceptionHandler;
   }
 
-  public void setSpanCustomizationInfo(SpanCustomizationInfo spanCustomizationInfo) {
+  public void setCoreSpanCustominzationInfoProvider(EventBasedStartSpanInfoProvider spanCustomizationInfo) {
```
Comment (sofiamorseletto) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/ProcessorRoute.java]: setCoreSpanCustomizationInfoProvider*

**Review by IvanAndresFritzler — APPROVED** [review]

**Pull request merged**